bindings/pyroot/pythonizations/python/ROOT/_pythonization/_function_utils.py

```diff
@@ -0,0 +1,224 @@
+                if t in map_tree_to_numbadeclares.keys(): # The column is a fundemental type from tree  
+                    return map_tree_to_numbadeclares[t]
+                elif '<' in  t: # The column type is a RVec<type>
+                    g = re.match('(.*)<(.*)>', t).groups(0)
+                    if g[1] in map_tree_to_numbadeclares.keys():
+                        return "RVec<" + map_tree_to_numbadeclares[g[1]] + ">"
+                    return "RVec<" + str(g[1]) + ">" # There are data type that leak into here. NOt sure from where. BUt need to implement something here.
+                    
+                else:
+                    return t
+            else :
+                return 'str'
+                #! Numba Declare does not support "string" type. Check _numbadeclare.Thus, Cannot pass string constants to the filter/Defines..
+        elif isinstance(x, bool):
+            return 'bool'
+        elif isinstance(x, int):
+            return 'int'
+        elif isinstance(x, float):
+            return 'float'
+        elif isinstance(x, np.ndarray): 
+            if x.dtype.type in  map_numpy_to_tree:
+                return "RVec<" + map_numpy_to_tree[x.dtype.type] + ">"
+            else:
+                raise TypeError(f"Support for {x.dtype.type} arrays is not yet supported.")
+        #! Need to work out how to map things like tuples, dicts, lists...
+        else:
+                raise TypeError(f"Type of {type(x)}:  {x} is not supported to be jitted.")
+            
+    def find_function_signature(self, func, funcargs):
+        """
+        Function to determine the function signature i.e the type of variables the function arguments and its corresponding value.
+        Args:
+	    `   func: Function whose signature is to be determined.
+                    P.S The function should not be crazy. It should have simple datatypes for the arguments and should not have default arguments. (Use funcargs instead)
+            funcargs: Dictionatry containing constant parameters the that func requires. Or any special mapping of the arguments of func.
+
+        Returns:
+            input_args: Map of func's input variables to values
+            type_args : Map of func'c input variables to its corresponding type.
+        """
+        try:
+            import inspect 
+        except:
+            raise ImportError("Failed to import inspect during call to determine function signature.")
+        func_sign = inspect.signature(func)
+        # Find the Return type
+        if func_sign.return_annotation is inspect._empty:
+            raise ValueError("Return type of the function is not mentioned.\n Function cannot be jitted as signature cannot be determined.")
+        else:
+            self.return_type = str(func_sign.return_annotation)
+        
+        # List of input parameters for function
+        self.params = list(func_sign.parameters.keys()) # ALl the input parameters the function needs.
```

Comment:
Nice, so you automatically get the names of the parameters of the function. Anyway, I believe the programming model should support two cases:
1. The user just runs `df.Filter(func)`
2. The user specifies the column names `df.Filter(func, col_names)`.

In 1. we just rely on the formal parameter names of the callable. In 2. the col_names specified by the user take precedence and we ignore the formal parameter names.